packages/core/src/components/widgets/FileWidget.tsx

```diff
@@ -81,81 +81,63 @@ function extractFileInfo(dataURLs: string[]) {
     });
 }
 
-type FileWidgetStateType = {
-  values: any[];
-  filesInfo: FileInfoType[];
-};
-
 /**
  *  The `FileWidget` is a widget for rendering file upload fields.
  *  It is typically used with a string property with data-url format.
  */
-class FileWidget<T, F> extends Component<
-  WidgetProps<T, F>,
-  FileWidgetStateType
-> {
-  constructor(props: WidgetProps<T, F>) {
-    super(props);
-    const { value } = props;
-    const values = Array.isArray(value) ? value : [value];
-    this.state = { values, filesInfo: extractFileInfo(values) };
-  }
-
-  shouldComponentUpdate(
-    nextProps: WidgetProps<T, F>,
-    nextState: FileWidgetStateType
-  ): boolean {
-    return shouldRender(this, nextProps, nextState);
-  }
+function FileWidget<T, F>({
+  multiple,
+  id,
+  readonly,
+  disabled,
+  onChange,
+  value,
+  autofocus = false,
+  options,
+}: WidgetProps<T, F>) {
+  const extractedFilesInfo = useMemo(
+    () =>
+      Array.isArray(value) ? extractFileInfo(value) : extractFileInfo([value]),
+    [value]
+  );
+  const [filesInfo, setFilesInfo] =
+    useState<FileInfoType[]>(extractedFilesInfo);
 
-  onChange = (event: ChangeEvent<HTMLInputElement>) => {
-    const { multiple, onChange } = this.props;
-    if (!event.target.files) {
-      return;
-    }
-    processFiles(event.target.files).then((filesInfo) => {
-      const state = {
-        values: filesInfo.map((fileInfo) => fileInfo.dataURL),
-        filesInfo,
-      };
-      this.setState(state, () => {
+  const handleChange = useCallback(
```

Comment:
Hey Jimmy, feel like making a pass through the fields and widgets to add `useCallback` in all the right places after this PR merges?

Reply:
sure thing!